### PR TITLE
#102 client  forgot password before logged in

### LINKED ISSUE #102: Client: Forgot password before logged in
The user can reset his password by clicking "forgot password" and entering his email

### PR DESCRIPTION
**Pull request by johannesfbraun**
closes #102 
frontend to send a reset link to the users mail
frontend to click on the resetPW link and send a new PW

### CODE REVIEW

**Comment by johannesfbraun** [comment]
closes #112 
frontend to send a reset link to the users mail
frontend to click on the resetPW link and send a new PW

File: client/src/ts/members/ForgotPassword.tsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
+/**
+ * Component for resetting the password by the user without the help of a admin, bofore logging in
+ *
```
Comment (LennartRukower) [inline client/src/ts/members/ForgotPassword.tsx:3]: Delete empty line.

File: client/src/ts/members/ForgotPassword.tsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    api.patch("/users/forgot-password", data
+    )
+      .then((res) => {
+        if (res.status === 200)
+          setSubmitBoolean(true);
+      });
+  };
+
+  /**
+   * check if input is an email
```
Comment (LennartRukower) [inline client/src/ts/members/ForgotPassword.tsx:85]: Start comment with a big letter.

File: client/src/ts/members/ForgotPassword.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+   */
+  const emailRegexTest = (testEmail: string) => {
+    const regex = /^.+@.+\..{1,6}/i;
+    return regex.test(testEmail);
+  };
+
+  /**
+   * return a warning if it is not a valid email
```
Comment (LennartRukower) [inline client/src/ts/members/ForgotPassword.tsx:93]: Start comment with a big letter.

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 4]
```diff
@@ -0,0 +1,207 @@
+/**
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:1]: It makes more sense to call this component/page "ResetPassword"

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      )
+        .then((res) => {
+          if (res.status === 200)
+            setResResponse200(true);
+        });
+    }
+  };
+
+  /**
+   * check if the new PW is okay: min 8 chars; mind 1 je num/a-z/A-Z
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:84]: Start comment with a big letter.

Comment (LennartRukower) [reply client/src/ts/members/ResetForgotPassword.tsx:84]: Please don't use abbreviation like PW.

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+   */
+  const checkNewPassword = (testString: string) => {
+    const regex = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{8,}/;
+    return regex.test(testString);
+  };
+
+  /**
+   * new password fild
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:92]: Start comment with a big letter.
This is not a sufficient comment for this function. What does this function do?

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+      return (
+        <Textfield className={classes.inputfield} id="newpassword" label="neues Passwort" type="password" value={newPassword}
+          onChange={event => { setNewPassword(event.target.value); }}
+          fullWidth />
+      );
+    }
+  };
+
+  /**
+   * new password field validation
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:111]: Start comment with a big letter.
This is not a sufficient comment for this function. What does this function do?

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+      return (
+        <Textfield error className={classes.inputfield} id="newpasswordvalidation" label="neues Passwort wiederholen" type="password" helperText="Die Passwörter müsen gleich sein" value={newPasswordValidation}
+          onChange={event => { setNewPasswordValidation(event.target.value); }}
+          fullWidth />
+      );
+    }
+  };
+
+  /**
+   * check if input is an email
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:130]: Start comment with a big letter.

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+   */
+  const emailRegexTest = (testEmail: string) => {
+    const regex = /^.+@.+\..{1,6}/i;
+    return regex.test(testEmail);
+  };
+
+  /**
+   * return a warning if it is not a valid email
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:138]: Start comment with a big letter.

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+  const invalidEmailWarning = () => {
+    if (!emailRegexTest(email)) {
+      return "This is not an Email";
+    } else {
+      return "";
+    }
+  };
+
+  /**
+   * if res is code 200
```
Comment (LennartRukower) [inline client/src/ts/members/ResetForgotPassword.tsx:149]: Start comment with a big letter.

File: server/src/members/membersController.ts [hunk 11]
```diff
@@ -98,6 +98,22 @@ export const changePassword = (req: Request, res: Response): void => {
     });
 };
 
+/**
+ * @param req email
+ * @param res 200 if process is available, 404 else
+ */
+export const sendPasswordResetLink = (req: Request, res:Response): void => {
+  res.status(200).send("not implemented");
```
Comment (LennartRukower) [inline server/src/members/membersController.ts:106]: Stubs in the backend should not be used.
Implement it in another branch and merge it before doing a pull request of the frontend.

@christophtrybek, your thoughts?

File: server/src/members/membersController.ts [hunk 12]
```diff
@@ -98,6 +98,22 @@ export const changePassword = (req: Request, res: Response): void => {
...
+};
+
+/**
+ * @param req email
+ * @param res 200 if process is available, 404 else
+ */
+ export const resetPasswordWithKey = (req: Request, res:Response): void => {
+  res.status(200).send("not implemented");
```
Comment (LennartRukower) [inline server/src/members/membersController.ts:114]: Stubs in the backend should not be used.
Implement it in another branch and merge it before doing a pull request of the frontend.

@christophtrybek, your thoughts?

File: server/src/members/membersController.ts [hunk 13]
```diff
@@ -98,6 +98,22 @@ export const changePassword = (req: Request, res: Response): void => {
     });
 };
 
+/**
+ * @param req email
```
Comment (LennartRukower) [inline server/src/members/membersController.ts:102]: This is not a sufficient comment for this function. What does this function do?

File: server/src/members/membersController.ts [hunk 14]
```diff
@@ -98,6 +98,22 @@ export const changePassword = (req: Request, res: Response): void => {
...
+ * @param res 200 if process is available, 404 else
+ */
+export const sendPasswordResetLink = (req: Request, res:Response): void => {
+  res.status(200).send("not implemented");
+};
+
+/**
+ * @param req email
```
Comment (LennartRukower) [inline server/src/members/membersController.ts:110]: This is not a sufficient comment for this function. What does this function do?

**Review by LennartRukower — CHANGES_REQUESTED** [review]
There are a few minor things I noticed.
I realized that some of these things (english names for routes and comments beginning with a small lettter) also occur in the ChangePassword-branch (#33), which was already merged.

@christophtrybek, what do you think is the best course of action? Should @johannesfbraun make a new issue and branch to "fix" these things?

**Comment by LennartRukower** [comment]
> closes #112
> frontend to send a reset link to the users mail
> frontend to click on the resetPW link and send a new PW

These things should be in the description of the pull request.

File: client/src/ts/members/ResetForgotPassword.tsx (hunk 4 above)
Comment (johannesfbraun) [reply client/src/ts/members/ResetForgotPassword.tsx:1]: I think the "Forgot" helps to clarify that it belongs to "ForgotPassword"

File: client/src/ts/members/ResetForgotPassword.tsx [hunk 15]
```diff
@@ -0,0 +1,207 @@
+/**
+ * Component for resetting the password by the user without the help of a admin, when the user gets the reset link
+ *
+ */
+import React, { useState, useContext } from "react";
```
Comment (qnewald) [inline client/src/ts/members/ResetForgotPassword.tsx:5]: "useContext" Remove unused imports

File: client/src/ts/members/ForgotPassword.tsx [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+  };
+
+  /**
+   * The button was clicked and the user is informed about the action
+   */
+  const informationOfSubmit = () => {
+    if (submitBoolean) {
+      return (
+        <Paper className={classes.paper}>
+          If you gave your correct email, you should recieve a new password
```
Comment (qnewald) [inline client/src/ts/members/ForgotPassword.tsx:109]: Mention that an email will be sent with instructions on how to reset the password.
For example: "If the entered email exists, we have sent you an email with instructions on how to reset your password."

**Review by qnewald — CHANGES_REQUESTED** [review]
When entering an E-Mail in the Forgot Password E-Mail field and hitting "send password" the page sometimes refreshes and the message that an email was sent if the email is correct will not be displayed.

**Pull request closed without merging**